### DISCUSSION TITLE
Google Cloud Batch executor with private IPs

### DISCUSSION THREAD

**Discussion by siddharthab** [nextflow-io/nextflow#5167] [General]
We discovered that if you use private IP addresses with Google Cloud Batch (Nextflow directive `batch.usePrivateAddress`), then, even if you have a Cloud Router and Cloud NAT set up in your VPC network, container images that are hosted outside of Google managed registries won't quite work. This applies to images hosted on quay.io, ghcr.io, Seqera Wave, etc.

This basically means that `batch.usePrivateAddress` should be used in only limited situations when you know that all your images are either coming from Google Artifact or Container Registry, or is one of the popular images that is transparently mirrored by Google Container Registry.

The cause for this is an engineering issue with their VM orchestration which tries to pull the image before the network stack is fully operational, and so that first connection takes 20-40 seconds, but the docker client installed on the disk image (`batch-cos`) has a hardcoded connection timeout of 15 seconds. You will see error messages like:
```
Error getting v2 registry: Get \"https://quay.io/v2/\": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

I have a support case open with them but not sure if there will be any resolution on it soon.

Sharing widely in case someone else has had similar problems or for any future users who might run into this error.

**Comment by bentsherman** [comment]
This is documented in our docs: https://nextflow.io/docs/latest/config.html#cloud-batch

Interesting to see that they might be able to solve it

**Reply by siddharthab** [reply]
Thank you for the reminder. I had forgotten about the documentation. I read through it when first setting up, but it was not clear if the limitations stated were because of the need for a NAT in the VPC. Upon setting up a NAT, things should technically work, but it seems like GCP has a bug on their end.

Maybe an additional line in the documentation that says this is the case even for networks with NAT, and that this issue has been reported to GCP?

**Reply by siddharthab** [reply]
Google has fixed the issue on their end. We just need to update the documentation here to say that if one is using private IPs, then they also need to set up a NAT component in their network.
